### PR TITLE
[feature] #2280: Produce permission events when role is granted/revoked

### LINKED ISSUE #2280: Notify event listners when a role is Revoked
## Problem
As noted in [PR #2264](https://github.com/hyperledger/iroha/pull/2264) an account can lose a permission that's included in a role and event listeners that look for the permission being removed will not be notified.
## Solution
 It would be nice if  event listeners that look for `PermissionAdded` / `PermissionRemoved` to also be notified if a role that contains this permission is being revoked.

### PR DESCRIPTION
**Pull request by Erigara**
Signed-off-by: Shanin Roman <shanin1000@yandex.ru>

### Description of the Change

- Add mechanism to return multiple events from `modify_*` functions;
- Fixed error when revoked role does not exist;
- Revoke/Grant role also produce `PermissionAdded/Removed` events.

### Issue

Closes #2280.

### Benefits

More detailed events.

### Possible Drawbacks

None.

### CODE REVIEW

**Review by SamHSmith — APPROVED** [review]
Approved. You may want to consider adding a test. Otherwise LGTM.

**Review by QuentinI — COMMENTED** [review]
Roles are also revoked when unregistering a role, so if we want to cover all such cases, we should generate token events [there](https://github.com/hyperledger/iroha/blob/iroha2-dev/core/src/smartcontracts/isi/world.rs#L178).

File: core/src/wsv.rs [hunk 1]
```diff
@@ -430,17 +430,31 @@ impl WorldStateView {
...
         f: impl FnOnce(&World) -> Result<WorldEvent, Error>,
     ) -> Result<(), Error> {
-        let world_event = f(&self.world)?;
-        let data_events: SmallVec<[DataEvent; 3]> = world_event.into();
+        self.modify_world_multiple_events(move |world| f(world).map(std::iter::once))
+    }
+
+    /// Get `World` and pass it to closure to modify it
+    ///
+    /// Produces events in the `WSV` that are produced by `f` during execution.
```
Comment (outoftardis) [inline core/src/wsv.rs:443]: "produces events ... that are produced" -- this should be rephrased

maybe like this (if I understood you correctly): "The function takes events from `WSV` that were produced during the execution of `f` and modifies them."?

File: core/src/wsv.rs [hunk 2]
```diff
@@ -625,58 +652,103 @@ impl WorldStateView {
...
     pub fn modify_asset(
         &self,
         id: &<Asset as Identifiable>::Id,
         f: impl FnOnce(&mut Asset) -> Result<AssetEvent, Error>,
     ) -> Result<(), Error> {
-        self.modify_account(&id.account_id, |account| {
+        self.modify_asset_multiple_events(id, move |asset| f(asset).map(std::iter::once))
+    }
+
+    /// Get `Asset` by its id
```
Comment (outoftardis) [inline core/src/wsv.rs:695]: function names start with `modify` but function descriptions are a bit different:

1) `and modify it with f`
2) `with an ability to modify it`
3) `and pass it to closure to modify it`
4) nothing

I'm trying to understand whether this is intentional: do all four cases mean different things? if not, then the descriptions should not be different

File: core/src/wsv.rs [hunk 3]
```diff
@@ -430,17 +430,31 @@ impl WorldStateView {
     /// Example: account events before domain events.
     ///
     /// # Errors
-    /// Fails if `f` fails
-    ///
-    /// # Panics
```
[the remark this answers is not in the document]
Comment (Erigara) [reply core/src/wsv.rs:435]: This section is outdated, as you see there is no usage of `self.events` at all.

File: core/src/wsv.rs [hunk 4]
```diff
@@ -430,17 +430,31 @@ impl WorldStateView {
...
+    /// # Errors
+    /// Forward errors from `f`
+    pub fn modify_world_multiple_events<I: IntoIterator<Item = WorldEvent>>(
+        &self,
+        f: impl FnOnce(&World) -> Result<I, Error>,
+    ) -> Result<(), Error> {
+        let world_events = f(&self.world)?;
+        let data_events: SmallVec<[DataEvent; 3]> = world_events
+            .into_iter()
+            .flat_map(SmallVec::<[DataEvent; 3]>::from)
```
[the remark this answers is not in the document]
Comment (Erigara) [reply core/src/wsv.rs:456]: Good point!

File: core/src/wsv.rs (hunk 2 above)
Comment (Erigara) [reply core/src/wsv.rs:695]: Use case of all this functions is the same.
We are passing object (world/asset/account/...) into closure `f` and receive bunch of events or error from it. 
We cant know what `f` is actually do with provided object (`f` can modify it or leave it as is).

File: core/src/wsv.rs [hunk 5]
```diff
@@ -625,58 +652,103 @@ impl WorldStateView {
...
-        self.modify_account(&id.account_id, |account| {
+        self.modify_asset_multiple_events(id, move |asset| f(asset).map(std::iter::once))
+    }
+
+    /// Get `Asset` by its id
+    ///
+    /// # Errors
+    /// - If there are no such asset or account
+    /// - Forward errors from `f`
+    #[allow(clippy::missing_panics_doc)]
```
Comment (Erigara) [inline core/src/wsv.rs:700]: Not relevant any more

**Comment by Erigara** [comment]
> Roles are also revoked when unregistering a role, so if we want to cover all such cases, we should generate token events [there](https://github.com/hyperledger/iroha/blob/iroha2-dev/core/src/smartcontracts/isi/world.rs#L178).

Good catch, thanks

File: core/src/wsv.rs (hunk 1 above)
Comment (Erigara) [reply core/src/wsv.rs:443]: > "produces events ... that are produced" -- this should be rephrased

Yeah, this sound confusing.

I would slightly change you suggestion: "The function put events produced during `f` execution into `event_buffer`"

File: core/src/wsv.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Erigara) [reply core/src/wsv.rs:456]: I have added function `WorldEvent::flatten(self)`.

File: core/src/wsv.rs [hunk 6]
```diff
@@ -523,21 +529,34 @@ impl WorldStateView {
         Ok(value)
     }
 
-    /// Get `Domain` and pass it to closure to modify it
+    /// The same as [`Self::modify_domain_multiple_events`] except closure `f` return single [`DomainEvent`].
```
Comment (outoftardis) [inline core/src/wsv.rs:532]:
```suggestion
    /// The same as [`Self::modify_domain_multiple_events`] except closure `f` returns a single [`DomainEvent`].
```

File: core/src/wsv.rs [hunk 7]
```diff
@@ -622,61 +641,108 @@ impl WorldStateView {
         Ok(f(account))
     }
 
-    /// Get `Account` and pass it to closure to modify it
+    /// The same as [`Self::modify_account_multiple_events`] except closure `f` return single [`AccountEvent`].
```
Comment (outoftardis) [inline core/src/wsv.rs:644]:
```suggestion
    /// The same as [`Self::modify_account_multiple_events`] except closure `f` returns a single [`AccountEvent`].
```

File: core/src/wsv.rs [hunk 8]
```diff
@@ -622,61 +641,108 @@ impl WorldStateView {
...
             let account = domain
                 .account_mut(id)
                 .ok_or_else(|| FindError::Account(id.clone()))?;
-            f(account).map(DomainEvent::Account)
+            f(account).map(|events| events.into_iter().map(DomainEvent::Account))
         })
     }
 
-    /// Get `Asset` by its id
+    /// The same as [`Self::modify_asset_multiple_events`] except closure `f` return single [`AssetEvent`].
```
Comment (outoftardis) [inline core/src/wsv.rs:674]:
```suggestion
    /// The same as [`Self::modify_asset_multiple_events`] except closure `f` returns a single [`AssetEvent`].
```

File: core/src/wsv.rs [hunk 9]
```diff
@@ -622,61 +641,108 @@ impl WorldStateView {
...
         &self,
         id: &<Asset as Identifiable>::Id,
         f: impl FnOnce(&mut Asset) -> Result<AssetEvent, Error>,
     ) -> Result<(), Error> {
-        self.modify_account(&id.account_id, |account| {
+        self.modify_asset_multiple_events(id, move |asset| f(asset).map(std::iter::once))
+    }
+
+    /// Get [`Asset`] and pass it to `closure` to modify it
+    /// If asset value hits 0 after modification asset is removed from the [`Account`]
```
Comment (outoftardis) [inline core/src/wsv.rs:687]:
```suggestion
    /// Get [`Asset`] and pass it to `closure` to modify it.
    /// If asset value hits 0 after modification, asset is removed from the [`Account`].
```

File: core/src/wsv.rs [hunk 10]
```diff
@@ -622,61 +641,108 @@ impl WorldStateView {
...
                 assert!(account.remove_asset(id).is_some());
             }
 
-            event_result.map(AccountEvent::Asset)
+            events_result.map(|events| events.into_iter().map(AccountEvent::Asset))
         })
     }
 
-    /// Get `AssetDefinitionEntry` with an ability to modify it.
+    /// The same as [`Self::modify_asset_definition_multiple_events`] except closure `f` return single [`AssetDefinitionEvent`].
```
Comment (outoftardis) [inline core/src/wsv.rs:714]:
```suggestion
    /// The same as [`Self::modify_asset_definition_multiple_events`] except closure `f` returns a single [`AssetDefinitionEvent`].
```

File: core/src/wsv.rs [hunk 11]
```diff
@@ -851,17 +929,30 @@ impl WorldStateView {
         self.events_buffer.borrow_mut().push(event.into());
     }
 
-    /// Get chain of validators and modify it with `f`
-    ///
-    /// Produces [`PermissionValidatorEvent`] from `f`
+    /// The same as [`Self::modify_validator_multiple_events`] except closure `f` return single [`PermissionValidatorEvent`].
```
Comment (outoftardis) [inline core/src/wsv.rs:932]:
```suggestion
    /// The same as [`Self::modify_validator_multiple_events`] except closure `f` returns a single [`PermissionValidatorEvent`].
```

File: data_model/src/events/data/events.rs [hunk 12]
```diff
@@ -560,6 +560,49 @@ pub enum WorldEvent {
     PermissionValidator(permission::PermissionValidatorEvent),
 }
 
+impl WorldEvent {
+    /// Unfold [`Self`] and return vector of [`Event`]s in expanding scope order: from specific to general.
```
Comment (outoftardis) [inline data_model/src/events/data/events.rs:564]:
```suggestion
    /// Unfold [`Self`] and return vector of [`Event`]s in the expanding scope order: from specific to general.
```

File: core/src/wsv.rs [hunk 13]
```diff
@@ -423,24 +419,34 @@ impl WorldStateView {
...
     pub fn modify_world(
         &self,
         f: impl FnOnce(&World) -> Result<WorldEvent, Error>,
     ) -> Result<(), Error> {
-        let world_event = f(&self.world)?;
-        let data_events: SmallVec<[DataEvent; 3]> = world_event.into();
+        self.modify_world_multiple_events(move |world| f(world).map(std::iter::once))
+    }
+
+    /// Get [`World`] and pass it to `closure` to modify it
```
Comment (outoftardis) [inline core/src/wsv.rs:433]:
```suggestion
    /// Get [`World`] and pass it to `closure` to modify it.
```

File: core/src/wsv.rs [hunk 14]
```diff
@@ -423,24 +419,34 @@ impl WorldStateView {
...
+    ///
+    /// The function puts events produced by `f` into `events_buffer`
```
Comment (outoftardis) [inline core/src/wsv.rs:435]:
```suggestion
    /// The function puts events produced by `f` into `events_buffer`.
```

**Review by outoftardis — APPROVED** [review]
two minor things, otherwise doc comments are ok

File: core/src/wsv.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuentinI) [reply core/src/wsv.rs:435]: I think this referred to [this](https://github.com/hyperledger/iroha/pull/2871/files#diff-b210796f83124da8c546d1f8ad43c2be94f98fd576a9e7018404a2760775dc92R455) access, which may still panic, although to my understanding that would indicate a bug and thus doesn't warrant a "Panics" section.

**Review by QuentinI — COMMENTED** [review]
LGTM

**Review by QuentinI — APPROVED** [review]

**Review by appetrosyan — APPROVED** [review]

**Pull request merged**